File: swift_reference/Views/TributePaywallView.swift
import SwiftUI
import RevenueCat

struct TributePaywallView: View {
    let store: StoreViewModel
    var contextTitle: String?
    var contextMessage: String?

    @Environment(\.dismiss) private var dismiss
    @State private var selectedPlan: PlanType = .yearly
    @State private var showContent: Bool = false
    @State private var purchaseSuccess: Bool = false

    private enum PlanType {
        case monthly, yearly
    }

    var body: some View {
        VStack(spacing: 0) {
            ScrollView {
                VStack(spacing: 24) {
                    headerSection

                    if let title = contextTitle {
                        contextSection(title: title, message: contextMessage)
                    }

                    planCards
                    featuresSection
                }
                .padding(.horizontal, 24)
                .padding(.top, 20)
                .padding(.bottom, 32)
            }

            bottomSection
        }
        .background(TributeColor.charcoal)
        .presentationDragIndicator(.visible)
        .presentationBackground(TributeColor.charcoal)
        .opacity(showContent ? 1 : 0)
        .onAppear {
            withAnimation(.easeOut(duration: 0.4)) {
                showContent = true
            }
        }
        .alert("Error", isPresented: .init(
            get: { store.error != nil },
            set: { if !$0 { store.error = nil } }
        )) {
            Button("OK") { store.error = nil }
        } message: {
            Text(store.error ?? "")
        }
        .onChange(of: store.isPremium) { _, isPremium in
            if isPremium {
                purchaseSuccess = true
                DispatchQueue.main.asyncAfter(deadline: .now() + 1.2) {
                    dismiss()
                }
            }
        }
        .sensoryFeedback(.success, trigger: purchaseSuccess)
    }

    private var headerSection: some View {
        VStack(spacing: 10) {
            ZStack {
                Circle()
                    .fill(
                        RadialGradient(
                            colors: [TributeColor.golden.opacity(0.2), TributeColor.golden.opacity(0.04)],
                            center: .center,
                            startRadius: 0,
                            endRadius: 40
                        )
                    )
                    .frame(width: 72, height: 72)

                Image(systemName: "crown.fill")
                    .font(.system(size: 28))
                    .foregroundStyle(TributeColor.golden)
            }

            Text("Tribute Pro")
                .font(.system(.title2, design: .serif, weight: .bold))
                .foregroundStyle(.primary)

            Text("Go deeper in your walk with God.")
                .font(.system(.subheadline, design: .serif))
                .foregroundStyle(.secondary)
                .multilineTextAlignment(.center)
        }
    }

    private func contextSection(title: String, message: String?) -> some View {
        VStack(spacing: 6) {
            Text(title)
                .font(.system(.subheadline, design: .serif, weight: .semibold))
                .foregroundStyle(TributeColor.golden)
            if let message {
                Text(message)
                    .font(.caption)
                    .foregroundStyle(.secondary)
                    .multilineTextAlignment(.center)
            }
        }
        .padding(14)
        .frame(maxWidth: .infinity)
        .background(TributeColor.golden.opacity(0.06))
        .clipShape(.rect(cornerRadius: 12))
        .overlay(
            RoundedRectangle(cornerRadius: 12)
                .strokeBorder(TributeColor.golden.opacity(0.15), lineWidth: 0.5)
        )
    }

    private var planCards: some View {
        HStack(spacing: 12) {
            if let monthly = store.monthlyPackage {
                planCard(
                    title: "Monthly",
                    price: monthly.storeProduct.localizedPriceString,
                    subtitle: "per month",
                    isSelected: selectedPlan == .monthly,
                    badge: nil
                ) {
                    selectedPlan = .monthly
                }
            }

            if let annual = store.annualPackage {
                let monthlyEquiv = annual.storeProduct.price / 12
                let formatter = NumberFormatter()
                let _ = { formatter.numberStyle = .currency; formatter.locale = annual.storeProduct.priceFormatter?.locale ?? .current }()
                let monthlyStr = formatter.string(from: monthlyEquiv as NSDecimalNumber) ?? ""

                planCard(
                    title: "Yearly",
                    price: annual.storeProduct.localizedPriceString,
                    subtitle: "\(monthlyStr)/mo",
                    isSelected: selectedPlan == .yearly,
                    badge: store.monthlySavingsText
                ) {
                    selectedPlan = .yearly
                }
            }
        }
    }

    private func planCard(
        title: String,
        price: String,
        subtitle: String,
        isSelected: Bool,
        badge: String?,
        action: @escaping () -> Void
    ) -> some View {
        Button(action: action) {
            VStack(spacing: 8) {
                if let badge {
                    Text(badge)
                        .font(.system(size: 10, weight: .bold))
                        .foregroundStyle(TributeColor.charcoal)
                        .padding(.horizontal, 8)
                        .padding(.vertical, 3)
                        .background(TributeColor.golden)
                        .clipShape(Capsule())
                } else {
                    Spacer().frame(height: 19)
                }

                Text(title)
                    .font(.caption.weight(.semibold))
                    .foregroundStyle(isSelected ? TributeColor.golden : .secondary)
                    .tracking(0.5)

                Text(price)
                    .font(.system(.title3, design: .rounded, weight: .bold))
                    .foregroundStyle(isSelected ? .primary : .secondary)

                Text(subtitle)
                    .font(.caption2)
                    .foregroundStyle(.secondary)
            }
            .frame(maxWidth: .infinity)
            .padding(.vertical, 16)
            .padding(.horizontal, 8)
            .background(isSelected ? TributeColor.golden.opacity(0.08) : TributeColor.cardBackground)
            .clipShape(.rect(cornerRadius: 14))
            .overlay(
                RoundedRectangle(cornerRadius: 14)
                    .strokeBorder(
                        isSelected ? TributeColor.golden.opacity(0.4) : TributeColor.cardBorder,
                        lineWidth: isSelected ? 1.5 : 0.5
                    )
            )
        }
        .buttonStyle(.plain)
    }

    private var featuresSection: some View {
        VStack(alignment: .leading, spacing: 12) {
            let features: [(icon: String, text: String)] = [
                ("infinity", "Unlimited habits"),
                ("shield.fill", "SOS temptation support"),
                ("chart.bar.fill", "Detailed analytics & insights"),
                ("text.quote", "Custom purpose statements"),
                ("calendar.badge.clock", "52-week Year in Tribute heatmap"),
                ("bell.badge.fill", "Smart reminders"),
            ]

            ForEach(features, id: \.text) { feature in
                HStack(spacing: 12) {
                    Image(systemName: feature.icon)
                        .font(.system(size: 14))
                        .foregroundStyle(TributeColor.golden)
                        .frame(width: 22)

                    Text(feature.text)
                        .font(.subheadline)
                        .foregroundStyle(TributeColor.softGold)
                }
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(16)
        .background(TributeColor.cardBackground)
        .clipShape(.rect(cornerRadius: 14))
        .overlay(
            RoundedRectangle(cornerRadius: 14)
                .strokeBorder(TributeColor.cardBorder, lineWidth: 0.5)
        )
    }

    private var bottomSection: some View {
        VStack(spacing: 12) {
            if purchaseSuccess {
                HStack(spacing: 8) {
                    Image(systemName: "checkmark.circle.fill")
                        .foregroundStyle(TributeColor.sage)
                    Text("Welcome to Tribute Pro")
                        .font(.system(.subheadline, design: .serif, weight: .semibold))
                        .foregroundStyle(TributeColor.sage)
                }
                .frame(maxWidth: .infinity)
                .padding(.vertical, 14)
                .transition(.opacity)
            } else {
                Button {
                    Task {
                        let pkg = selectedPlan == .yearly ? store.annualPackage : store.monthlyPackage
                        if let pkg { await store.purchase(package: pkg) }
                    }
                } label: {
                    Group {
                        if store.isPurchasing {
                            ProgressView()
                                .tint(TributeColor.charcoal)
                        } else {
                            Text("Continue")
                        }
                    }
                    .tributeButton()
                }
                .disabled(store.isPurchasing || store.isLoading)

                HStack(spacing: 16) {
                    Button {
                        Task { await store.restore() }
                    } label: {
                        Text("Restore Purchases")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }

                    Button { dismiss() } label: {
                        Text("Not now")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }
                }
            }
        }
        .padding(.horizontal, 24)
        .padding(.bottom, 32)
        .padding(.top, 12)
    }
}

File: lib/views/shared/tribute_paywall_view.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../../providers/store_provider.dart';
import '../../theme/app_theme.dart';

class TributePaywallView extends StatefulWidget {
  final String? contextTitle;
  final String? contextMessage;

  const TributePaywallView({super.key, this.contextTitle, this.contextMessage});

  @override
  State<TributePaywallView> createState() => _TributePaywallViewState();
}

class _TributePaywallViewState extends State<TributePaywallView> {
  bool _yearlySelected = true;
  bool _purchaseSuccess = false;

  static const _features = [
    (Icons.all_inclusive_rounded, 'Unlimited habits'),
    (Icons.shield_rounded, 'SOS temptation support'),
    (Icons.bar_chart_rounded, 'Detailed analytics & insights'),
    (Icons.format_quote_rounded, 'Custom purpose statements'),
    (Icons.calendar_month_rounded, '52-week Year in Tribute heatmap'),
    (Icons.notifications_rounded, 'Smart reminders'),
  ];

  @override
  Widget build(BuildContext context) {
    final store = context.watch<StoreProvider>();

    if (store.isPremium && !_purchaseSuccess) {
      _purchaseSuccess = true;
      WidgetsBinding.instance.addPostFrameCallback((_) {
        if (!mounted) return;
        setState(() {});
        Future.delayed(const Duration(milliseconds: 1200), () {
          if (mounted) Navigator.pop(context);
        });
      });
    }

    return Scaffold(
      backgroundColor: TributeColor.charcoal,
      body: SafeArea(
        child: Column(children: [
          Expanded(
            child: SingleChildScrollView(
              padding: const EdgeInsets.fromLTRB(24, 20, 24, 32),
              child: Column(children: [
                _headerSection(),
                if (widget.contextTitle != null) ...[
                  const SizedBox(height: 20),
                  _contextSection(),
                ],
                const SizedBox(height: 24),
                _planCards(store),
                const SizedBox(height: 20),
                _featuresSection(),
              ]),
            ),
          ),
          _bottomSection(store),
        ]),
      ),
    );
  }

  Widget _headerSection() {
    return Column(children: [
      Container(
        width: 72, height: 72,
        decoration: BoxDecoration(
          shape: BoxShape.circle,
          gradient: RadialGradient(colors: [
            TributeColor.golden.withValues(alpha: 0.2),
            TributeColor.golden.withValues(alpha: 0.04),
          ]),
        ),
        child: const Icon(Icons.workspace_premium_rounded, size: 28, color: TributeColor.golden),
      ),
      const SizedBox(height: 10),
      const Text('Tribute Pro',
          style: TextStyle(fontSize: 24, fontWeight: FontWeight.w700, color: TributeColor.warmWhite)),
      const SizedBox(height: 6),
      Text('Go deeper in your walk with God.',
          textAlign: TextAlign.center,
          style: TextStyle(fontSize: 15, color: Colors.white.withValues(alpha: 0.6))),
    ]);
  }

  Widget _contextSection() {
    return Container(
      width: double.infinity,
      padding: const EdgeInsets.all(14),
      decoration: BoxDecoration(
        color: TributeColor.golden.withValues(alpha: 0.06),
        borderRadius: BorderRadius.circular(12),
        border: Border.all(color: TributeColor.golden.withValues(alpha: 0.15), width: 0.5),
      ),
      child: Column(children: [
        Text(widget.contextTitle!,
            textAlign: TextAlign.center,
            style: const TextStyle(fontSize: 14, fontWeight: FontWeight.w600, color: TributeColor.golden)),
        if (widget.contextMessage != null) ...[
          const SizedBox(height: 4),
          Text(widget.contextMessage!,
              textAlign: TextAlign.center,
              style: TextStyle(fontSize: 12, color: Colors.white.withValues(alpha: 0.5))),
        ],
      ]),
    );
  }

  Widget _planCards(StoreProvider store) {
    final monthly = store.monthlyPackage;
    final annual = store.annualPackage;
    return Row(children: [
      if (monthly != null) ...[
        Expanded(child: _planCard(
          title: 'Monthly',
          price: '\$${monthly.storeProduct.price.toStringAsFixed(2)}',
          subtitle: 'per month',
          selected: !_yearlySelected,
          badge: null,
          onTap: () => setState(() => _yearlySelected = false),
        )),
        const SizedBox(width: 12),
      ],
      if (annual != null)
        Expanded(child: _planCard(
          title: 'Yearly',
          price: '\$${annual.storeProduct.price.toStringAsFixed(2)}',
          subtitle: 'best value',
          selected: _yearlySelected,
          badge: store.monthlySavingsText,
          onTap: () => setState(() => _yearlySelected = true),
        ))
      else if (monthly == null)
        _noPlansFallback(),
    ]);
  }

  Widget _planCard({
    required String title,
    required String price,
    required String subtitle,
    required bool selected,
    required String? badge,
    required VoidCallback onTap,
  }) {
    return GestureDetector(
      onTap: onTap,
      child: Container(
        padding: const EdgeInsets.fromLTRB(8, 16, 8, 16),
        decoration: BoxDecoration(
          color: selected ? TributeColor.golden.withValues(alpha: 0.08) : TributeColor.cardBackground,
          borderRadius: BorderRadius.circular(14),
          border: Border.all(
            color: selected ? TributeColor.golden.withValues(alpha: 0.4) : TributeColor.cardBorder,
            width: selected ? 1.5 : 0.5,
          ),
        ),
        child: Column(children: [
          if (badge != null)
            Container(
              padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 3),
              decoration: BoxDecoration(
                color: TributeColor.golden,
                borderRadius: BorderRadius.circular(20),
              ),
              child: Text(badge,
                  style: const TextStyle(fontSize: 10, fontWeight: FontWeight.w700, color: TributeColor.charcoal)),
            )
          else
            const SizedBox(height: 19),
          const SizedBox(height: 8),
          Text(title,
              style: TextStyle(fontSize: 11, fontWeight: FontWeight.w600,
                  color: selected ? TributeColor.golden : Colors.white.withValues(alpha: 0.5),
                  letterSpacing: 0.5)),
          const SizedBox(height: 4),
          Text(price,
              style: TextStyle(fontSize: 20, fontWeight: FontWeight.w700,
                  color: selected ? TributeColor.warmWhite : Colors.white.withValues(alpha: 0.5))),
          const SizedBox(height: 2),
          Text(subtitle,
              style: TextStyle(fontSize: 10, color: Colors.white.withValues(alpha: 0.4))),
        ]),
      ),
    );
  }

  Widget _noPlansFallback() {
    return Expanded(
      child: Center(
        child: Text('Loading plans...',
            style: TextStyle(fontSize: 13, color: Colors.white.withValues(alpha: 0.4))),
      ),
    );
  }

  Widget _featuresSection() {
    return Container(
      padding: const EdgeInsets.all(16),
      decoration: TributeDecorations.card,
      child: Column(
        children: _features.map((f) => Padding(
              padding: const EdgeInsets.only(bottom: 12),
              child: Row(children: [
                Icon(f.$1, size: 16, color: TributeColor.golden),
                const SizedBox(width: 12),
                Text(f.$2, style: const TextStyle(fontSize: 14, color: TributeColor.softGold)),
              ]),
            )).toList(),
      ),
    );
  }

  Widget _bottomSection(StoreProvider store) {
    return Padding(
      padding: const EdgeInsets.fromLTRB(24, 12, 24, 32),
      child: Column(children: [
        if (_purchaseSuccess)
          Row(mainAxisAlignment: MainAxisAlignment.center, children: [
            const Icon(Icons.check_circle_rounded, color: TributeColor.sage, size: 18),
            const SizedBox(width: 8),
            const Text('Welcome to Tribute Pro',
                style: TextStyle(fontSize: 15, fontWeight: FontWeight.w600, color: TributeColor.sage)),
          ])
        else ...[
          SizedBox(
            width: double.infinity,
            child: ElevatedButton(
              onPressed: (store.isPurchasing || store.isLoading) ? null : () => _purchase(store),
              style: ElevatedButton.styleFrom(
                backgroundColor: TributeColor.golden,
                foregroundColor: TributeColor.charcoal,
                padding: const EdgeInsets.symmetric(vertical: 16),
                shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(14)),
              ),
              child: store.isPurchasing
                  ? const SizedBox(width: 20, height: 20,
                      child: CircularProgressIndicator(strokeWidth: 2, color: TributeColor.charcoal))
                  : const Text('Continue', style: TextStyle(fontWeight: FontWeight.w600, fontSize: 16)),
            ),
          ),
          const SizedBox(height: 12),
          Row(mainAxisAlignment: MainAxisAlignment.center, children: [
            TextButton(
              onPressed: store.isLoading ? null : () => store.restore(),
              child: Text('Restore Purchases',
                  style: TextStyle(fontSize: 12, color: Colors.white.withValues(alpha: 0.5))),
            ),
            Text('\u00B7', style: TextStyle(color: Colors.white.withValues(alpha: 0.3))),
            TextButton(
              onPressed: () => Navigator.pop(context),
              child: Text('Not now',
                  style: TextStyle(fontSize: 12, color: Colors.white.withValues(alpha: 0.5))),
            ),
          ]),
          if (store.error != null)
            Padding(
              padding: const EdgeInsets.only(top: 8),
              child: Text(store.error!,
                  textAlign: TextAlign.center,
                  style: const TextStyle(fontSize: 12, color: TributeColor.warmCoral)),
            ),
        ],
      ]),
    );
  }

  Future<void> _purchase(StoreProvider store) async {
    final pkg = _yearlySelected ? store.annualPackage : store.monthlyPackage;
    if (pkg != null) await store.purchase(pkg);
  }
}
